File: ReactorCore.py
#Core simulation logic

from Old Stuff.equations import (
    calc_k_eff,
    calc_reactivity,
    evolve_flux,
    calc_power_output,
    calc_coolant_removal,
    calc_temp_change
)
from Old Stuff.Config import DEFAULT_CORE_HEAT_CAPACITY, DEFAULT_INITIAL_TEMP, DEFAULT_INITIAL_FLUX, MELTDOWN_TEMP, SCRAM_REACTIVITY

class ReactorCore:
    def __init__(self, fuel_type: str, enrichment: float, rod_percent: float, coolant_rate: float):
        self.fuel_type = fuel_type
        self.enrichment = enrichment
        self.rod_percent = rod_percent
        self.coolant_rate = coolant_rate

        self.time = 0
        self.temperature = DEFAULT_INITIAL_TEMP
        self.flux = DEFAULT_INITIAL_FLUX
        self.power_output = 0
        self.reactivity = 0
        self.status = "SAFE"

    def update_state(self, dt: float):
        """Advance simulation by dt seconds."""
        k_eff = calc_k_eff(self.fuel_type, self.enrichment, self.rod_percent)
        self.reactivity = calc_reactivity(k_eff)
        self.flux = evolve_flux(self.flux, self.reactivity)
        self.power_output = calc_power_output(self.flux, self.enrichment)

        q_cool = calc_coolant_removal(self.temperature, self.coolant_rate)
        delta_temp = calc_temp_change(self.power_output, q_cool, DEFAULT_CORE_HEAT_CAPACITY, dt)
        self.temperature += delta_temp

        self.time += dt
        self.check_status()

    def check_status(self):
        """Update reactor status based on safety limits."""
        if self.temperature > MELTDOWN_TEMP:
            self.status = "MELTDOWN"
        elif self.reactivity > SCRAM_REACTIVITY:
            self.status = "SCRAM"
        elif self.temperature > 900:
            self.status = "WARNING"
        else:
            self.status = "SAFE"

    def to_dict(self):
        """Return a snapshot of current state."""
        return {
            "time": self.time,
            "temperature": self.temperature,
            "flux": self.flux,
            "power_output": self.power_output,
            "reactivity": self.reactivity,
            "status": self.status,
            "rod_percent": self.rod_percent,
            "coolant_rate": self.coolant_rate,
            "enrichment": self.enrichment,
            "fuel_type": self.fuel_type
        }

File: web/js/ReactorCore.js
/**
 * CoreSim — Reactor Core State Machine
 *
 * Holds all instantaneous state for the reactor and advances
 * it forward by one timestep using the equations module.
 */

import {
    calcKEff,
    calcReactivity,
    evolveFlux,
    calcStartupRate,
    calcPowerOutput,
    calcCoolantRemoval,
    calcTempChange,
    calcCoolantOutletTemp,
    calcPressure,
    calcWaterLevel,
    calcRadiationLevel,
    calcContainmentPressure,
    calcEnergyOutput,
    PHYSICS
} from './equations.js';

export class ReactorCore {
    constructor(fuelType, enrichment, rodPercent, coolantRate) {
        // Configuration (can change at runtime via controls)
        this.fuelType = fuelType;
        this.enrichment = enrichment;
        this.rodPercent = rodPercent;
        this.coolantRate = coolantRate;

        // ── Simulation clock ──
        this.time = 0;

        // ── Nuclear state ──
        this.kEff = 1.0;
        this.reactivity = 0;
        this.flux = 1e14;          // n/cm²·s — typical PWR operating flux
        this.prevFlux = 1e14;
        this.startupRate = 0;      // decades/min (SUR)

        // ── Thermal state ──
        this.fuelTemp = PHYSICS.DEFAULT_INLET_TEMP;      // °C
        this.coolantTemp = PHYSICS.DEFAULT_INLET_TEMP;    // °C (outlet)
        this.powerOutput = 0;      // Watts

        // ── Hydraulic state ──
        this.pressure = PHYSICS.NOMINAL_PRESSURE;         // MPa
        this.waterLevel = PHYSICS.NOMINAL_WATER_LEVEL;    // %

        // ── Environmental state ──
        this.radiationLevel = 0;   // mSv/hr
        this.containmentPressure = PHYSICS.CONTAINMENT_BASE_PRESSURE; // MPa
        this.energyOutputMWh = 0;  // cumulative MWh

        // ── Status ──
        this.status = "OFFLINE";
        this.scramActive = false;
    }

    /**
     * Advance the simulation by dt seconds.
     */
    updateState(dt) {
        // 1. Neutronics
        this.kEff = calcKEff(this.fuelType, this.enrichment, this.rodPercent, this.fuelTemp);
        this.reactivity = calcReactivity(this.kEff);

        this.prevFlux = this.flux;
        this.flux = evolveFlux(this.flux, this.reactivity, dt);

        // Clamp flux to avoid numerical blow-up
        this.flux = Math.max(1e2, Math.min(this.flux, 1e20));

        this.startupRate = calcStartupRate(this.prevFlux, this.flux, dt);

        // 2. Thermal
        this.powerOutput = calcPowerOutput(this.flux, this.enrichment);
        const qCool = calcCoolantRemoval(this.fuelTemp, this.coolantRate);
        const deltaTemp = calcTempChange(this.powerOutput, qCool, dt);
        this.fuelTemp += deltaTemp;

        // Clamp to physical range
        if (this.fuelTemp < PHYSICS.DEFAULT_INLET_TEMP) {
            this.fuelTemp = PHYSICS.DEFAULT_INLET_TEMP;
        }

        this.coolantTemp = calcCoolantOutletTemp(
            PHYSICS.DEFAULT_INLET_TEMP,
            this.powerOutput,
            this.coolantRate
        );

        // 3. Hydraulic
        this.pressure = calcPressure(this.coolantTemp);
        this.waterLevel = calcWaterLevel(this.pressure);

        // 4. Environmental
        this.radiationLevel = calcRadiationLevel(this.flux, this.powerOutput);
        this.containmentPressure = calcContainmentPressure(this.fuelTemp);
        this.energyOutputMWh = calcEnergyOutput(this.energyOutputMWh, this.powerOutput, dt);

        // 5. Clock & status
        this.time += dt;
        this.checkStatus();
    }

    checkStatus() {
        if (this.fuelTemp > PHYSICS.MELTDOWN_TEMP) {
            this.status = "MELTDOWN";
        } else if (this.scramActive) {
            this.status = "SCRAM";
        } else if (this.fuelTemp > PHYSICS.SCRAM_TEMP) {
            // Auto-SCRAM on extreme temperature
            this.triggerScram();
            this.status = "SCRAM";
        } else if (this.fuelTemp > PHYSICS.WARNING_TEMP) {
            this.status = "WARNING";
        } else {
            this.status = "SAFE";
        }
    }

    triggerScram() {
        this.scramActive = true;
        this.rodPercent = 100;
    }

    clearScram() {
        this.scramActive = false;
    }

    // ── Control setters ──
    setRodPercent(val) {
        this.rodPercent = val;
        // If user manually moves rods off 100%, clear SCRAM
        if (val < 100) this.scramActive = false;
    }
    setCoolantRate(val) { this.coolantRate = val; }

    /**
     * Return a plain-object snapshot of the full reactor state.
     * This is what gets pushed into history[] and sent to the UI.
     */
    getSnapshot() {
        return {
            time: this.time,

            // Nuclear
            kEff: this.kEff,
            reactivity: this.reactivity,
            flux: this.flux,
            startupRate: this.startupRate,
            rodPercent: this.rodPercent,

            // Thermal
            fuelTemp: this.fuelTemp,
            coolantTemp: this.coolantTemp,
            powerOutput: this.powerOutput,
            powerMW: this.powerOutput / 1e6,

            // Hydraulic
            pressure: this.pressure,
            waterLevel: this.waterLevel,
            coolantRate: this.coolantRate,

            // Environmental
            radiationLevel: this.radiationLevel,
            containmentPressure: this.containmentPressure,
            energyOutputMWh: this.energyOutputMWh,

            // Status
            status: this.status,
            scramActive: this.scramActive,
        };
    }
}
